### PR TITLE
iscsi: don't write json medata file when the volume is already mounted.

### LINKED ISSUE #91736: Can't run two pods with read-only iSCSI volumes
**What happened**:
Running two pods that use the same iSCSI volume, one of them fails with:

```
Warning  FailedMount  6s (x5 over 14s)  kubelet, 127.0.0.1  MountVolume.WaitForAttach failed for volume "test-volume" : iscsi: create /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/127.0.0.1:3260-iqn.2003-01.io.k8s:jsafrane-
lun-0/iscsi.json err open /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/127.0.0.1:3260-iqn.2003-01.io.k8s:jsafrane-lun-0/iscsi.json: read-only file system
```

**What you expected to happen**:
Both pods should be running

**Anything else we need to know?**:
It's caused by iscsi volume plugin persiting iSCSI settings into the mount directory itself (e.g. `/var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/127.0.0.1:3260-iqn.2003-01.io.k8s:jsafrane-lun-0`):

1. The volume plugin creates the global mount directory.
2. The volume plugin creates iscsi.json there.
3. The volume plugin mounts the volume there. This shadows the iscsi.json file, but it's fine, as long as no other pod uses the same volume on the same node.
4. User creates a second pod that uses the volume.
5. The volume plugin tries to create iscsi.json in the global volume directory. Since the volume is mounted, it will either writes the file to the volume (!) or fails, if the volume is read-only.

### PR DESCRIPTION
**Pull request by jsafrane**
iSCSI volume plugin persists volume metadata into global mount directory, before it is mounted. Content of the directory is shadowed by the volume mount. Therefore kubelet should not write metadata to the directory when a second pod uses the same volume on the same node.
    
1. The metadata were already persisted before mounting the volume for the first pod.
2. The global mount directory has the volume mounted, so any write there would write to the volume, which is undesirable.

Fixes: #91736

```release-note
Fixed using of a read-only iSCSI volume in multiple pods.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by jsafrane** [comment]
Can't find what's wrong with `Pre-provisioned PV (default fs)] subPath should support existing directories when readOnly specified in the volumeSource ` - the same test passes for in-line volume. And on my cluster.

**Comment by jsafrane** [comment]
I think I know what's going on: kernel iSCSI target looses data.  In kubelet logs I can see that the faulty test creates a pod with subPath, that gets correctly created, but them it run she same pod with the same subPath as read-only. Kubelet should not create the subpath dir (it already exists from the old pod), however, this subdir is not present on the volume.

Similarly, "volumes should store data" tests fails for iscsi from time to time, when it misses one write() from a pod that created test data on the volume few seconds before.

https://bugzilla.redhat.com/show_bug.cgi?id=1721218

We should rework the tests for some more stable iSCSI target implementation.

File: pkg/volume/iscsi/iscsi_util.go [hunk 1]
```diff
@@ -623,13 +654,22 @@ func (util *ISCSIUtil) DetachDisk(c iscsiDiskUnmounter, mntPath string) error {
...
 		return nil
 	}
 
 	err = util.detachISCSIDisk(c.exec, portals, iqn, iface, volName, initiatorName, found)
 	if err != nil {
 		return fmt.Errorf("failed to finish detachISCSIDisk, err: %v", err)
 	}
+	err = util.removeISCSI(mntPath)
+	if err != nil {
+		return fmt.Errorf("failed to remove metadata file: %s", err)
```
Comment (bertinatto) [inline pkg/volume/iscsi/iscsi_util.go:670]: nit: `%s` while (it seems like) existing code uses `%v`

File: pkg/volume/iscsi/iscsi_util.go [hunk 2]
```diff
@@ -204,34 +204,63 @@ func (util *ISCSIUtil) MakeGlobalVDPDName(iscsi iscsiDisk) string {
...
-	fp, err := os.Open(file)
-	if err != nil {
-		return fmt.Errorf("iscsi: open %s err %s", file, err)
+	jsonFilePath := getISCSIPersistFile(mnt)
+	fp, openerr := os.Open(jsonFilePath)
+	if openerr != nil {
+		if !os.IsNotExist(openerr) {
+			return fmt.Errorf("iscsi: failed to open %s: %s", jsonFilePath, openerr)
+		}
+		// JSON file does not exist. For backward compatibility: try iscsi.json in the mount directory.
```
Comment (bertinatto) [inline pkg/volume/iscsi/iscsi_util.go:227]: For how long do we want to be backward-compatible with `iscsi.json`? Can we add a TODO to remove this v1.2x?

**Comment by bertinatto** [comment]
Changes LGTM; will let other people review as well.

CC @j-griffith

File: pkg/volume/iscsi/iscsi_util.go [hunk 3]
```diff
@@ -204,34 +204,63 @@ func (util *ISCSIUtil) MakeGlobalVDPDName(iscsi iscsiDisk) string {
...
+		file := filepath.Join(mnt, "iscsi.json")
+		var err error
+		fp, err = os.Open(file)
+		if err != nil {
+			// Report original openerror
+			return fmt.Errorf("iscsi: failed to open %s: %s", file, openerr)
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/volume/iscsi/iscsi_util.go:233]: It's explicitly commented that `openerr` is returned. The error from the "new" file is IMO more interesting than the legacy one.

File: pkg/volume/iscsi/iscsi_util.go (hunk 2 above)
Comment (jsafrane) [reply pkg/volume/iscsi/iscsi_util.go:227]: TODO added

File: pkg/volume/iscsi/iscsi_util.go (hunk 1 above)
Comment (jsafrane) [reply pkg/volume/iscsi/iscsi_util.go]: It's really mixed in the code, but I added `%v` to couple of places.

**Comment by bswartz** [comment]
I have a high level question about this fix. Why can't we use a different mechanism to avoid reading/writing the json file in the case when the volume is already attached to the current node?

It seems like if we knew the volume was attached already when attaching to the second pod, we could skip the writing of the json file. And if we knew the volume was still attached to another pod while detaching from a pod, we couid skip the reading of the file. The contents of the file should only be necessary to do the final detach, and in that case the json file should never be shadowed by a mount.

I might be missing something about what we use the file contents for, but my understanding was that we only needed it to perform the logout, and we had a different way to do reference counting so we can reliably tell if we're not the first pod to attach to a given volume, and if we're not the last pod to detach from a given volume.

**Comment by jsafrane** [comment]
> The contents of the file should only be necessary to do the final detach, and in that case the json file should never be shadowed by a mount.

@bswartz, you're correct! I reworked the patch not to move the location of iSCSI metadata file and write it only for the first time. Please check the new PR descripion & code.

**Review by bswartz — APPROVED** [review]
/lgtm

**Comment by jsafrane** [comment]
/assign @andrewsykim @oomichi 
for framework change approval

**Comment by oomichi** [comment]
e2e framework part also seems good for me.

/approve

**Pull request merged**